pss/crypto/crypto.go

```diff
@@ -0,0 +1,656 @@
+	Payload   []byte // Parsed and validated message content
+	Raw       []byte // Unparsed but decrypted message content
+	Signature []byte // Signature from the sender
+	Salt      []byte // Salt used for validation
+	Padding   []byte // Padding applied
+
+	Src *ecdsa.PublicKey // Source public key used for signing the message
+	Dst *ecdsa.PublicKey // Destination public key used for encrypting the msessage asymmetrically
+
+	crypto *defaultCryptoBackend
+}
+
+func (msg receivedMessage) GetSrc() *ecdsa.PublicKey {
+	return msg.Src
+}
+
+// ValidateAndParse checks that the format and the signnature are correct. It also set Payload as the parsed message
+func (msg *receivedMessage) ValidateAndParse() bool {
+	end := len(msg.Raw)
+	if end < 1 {
+		return false
+	}
+	if isMessageSigned(msg.Raw[0]) {
+		end -= signatureLength
+		if end <= 1 {
+			return false
+		}
+		msg.Signature = msg.Raw[end : end+signatureLength]
+		msg.Src = msg.sigToPubKey()
+		if msg.Src == nil {
+			return false
+		}
+	}
+
+	beg := 1
+	payloadSize := 0
+	sizeOfPayloadSizeField := int(msg.Raw[0] & SizeMask) // number of bytes indicating the size of payload
+	if sizeOfPayloadSizeField != 0 {
+		payloadSize = int(bytesToUintLittleEndian(msg.Raw[beg : beg+sizeOfPayloadSizeField]))
+		if payloadSize+1 > end {
+			return false
+		}
+		beg += sizeOfPayloadSizeField
+		msg.Payload = msg.Raw[beg : beg+payloadSize]
+	}
+
+	beg += payloadSize
+	msg.Padding = msg.Raw[beg:end]
+	return true
+}
+
+// GetPayload obtains the parsed payload of the message. Should be called after ValidateAndParse
+func (msg receivedMessage) GetPayload() []byte {
+	return msg.Payload
+}
+
+// sigToPubKey returns the public key associated to the message's signature.
+// should only be called if message has been signed
+func (msg *receivedMessage) sigToPubKey() *ecdsa.PublicKey {
+	defer func() { recover() }() // in case of invalid signature
+
+	signedBytes := msg.Raw
+	if isMessageSigned(msg.Raw[0]) {
+		sz := len(msg.Raw) - signatureLength
+		signedBytes = msg.Raw[:sz]
+	}
+	pub, err := msg.crypto.sigToPub(signedBytes, msg.Signature)
+	if err != nil {
+		log.Error("failed to recover public key from signature", "err", err)
+		return nil
+	}
+	return pub
+}
+
+func newReceivedMessage(decrypted []byte, salt []byte, crypto *defaultCryptoBackend) *receivedMessage {
+	return &receivedMessage{
+		Raw:    decrypted,
+		Salt:   salt,
+		crypto: crypto,
+	}
+}
+
+func New() Crypto {
+	cryptoBackend = defaultCryptoBackend{
+		symKeys:     make(map[string][]byte),
+		privateKeys: make(map[string]*ecdsa.PrivateKey),
+	}
+	return &cryptoBackend
+}
+
+func NewCryptoUtils() CryptoUtils {
+	if cryptoBackend.privateKeys == nil {
+		New()
+	}
+	return &cryptoBackend
+}
+
+// == Message Crypto ==
+
+func (crypto *defaultCryptoBackend) WrapMessage(payload []byte, params *WrapParams) (data []byte, err error) {
+	var padding []byte
+	if createPadding {
+		padding, err = crypto.generateSecureRandomData(defaultPaddingByteSize)
```

Comment:
As far as I understand from this, the same amount of padding is always just added to the message.

I would prefer if messages were always rounded up to a multiple of `paddedSize` instead. The intention is to make protocol profiling more difficult by hiding the sizes of the payloads, but if you know the sizes are in reality "payload minus 16" or whatever, nothing much has been gained.

I would also prefer that this multiple is customizable by the sender's node, _and_ that (parts of) the padding can be arbitarily specified per message through an API (enabling the possiblity of embedding content in the padding itself).

(The changes described in the last paragraph can be deferred to a later PR.)

Reply:
Right now we have a fixed padding, yes. I have copied appendpadding -> pad method to really add a variable padding, although I'm not using it now. We can leave the padding details for a specific PR if you want.

Reply:
Could I beg you to create an issue for this?

Reply:
Of course